### PR TITLE
Feature/storage/quick query bug

### LINKED ISSUE #13441: [BUG] Run BlockBlobClient.QueryAsync() on a big csv file, then write the output stream to a file fail.
**Describe the bug**
Run BlockBlobClient.QueryAsync() on a big csv file (~230MB), then write the output stream to a file fail with  "Destination array was not long enough. Check the destination index, length, and the array's lower bounds."

And it seems the “ProgressHandler” won’t be called until the request finish, and start to read from output stream. I expect it will be invoke on download the result from server. Any comments?

**Expected behavior**
write the stream to a local file should success.

**Actual behavior (include Exception or Stack Trace)**
Failed with
```
System.ArgumentException: Destination array was not long enough. Check the destination index, length, and the array's lower bounds.
Parameter name: destinationArray
   at System.Array.Copy(Array sourceArray, Int32 sourceIndex, Array destinationArray, Int32 destinationIndex, Int32 length, Boolean reliable)
   at Azure.Storage.Blobs.BlobQuickQueryStream.ReadInternal(Boolean async, Byte[] buffer, Int32 offset, Int32 count)
   at Azure.Storage.Blobs.BlobQuickQueryStream.Read(Byte[] buffer, Int32 offset, Int32 count)
   at System.IO.Stream.CopyTo(Stream destination, Int32 bufferSize)
   at testTracking.Program.Main(String[] args) in C:\Users\weiwei\source\repos\testTracking\testTracking\Program.cs:line 74
```

**To Reproduce**
Steps to reproduce the behavior (include a code snippet, screenshot, or any additional information that might help us reproduce the issue)

Following is my repro program, I will share my test csv file Uri by mail.
The failure is on line `reader.CopyTo(fs); `

```
           string bloburi = "https://xxx.blob.core.windows.net/weiqatest/testcsv.csv?sv=2019-12-12&se=2020-12-29T16%3A00%3A00Z&sr=b&sp=r&sig=xxx";
           BlockBlobClient blob = new BlockBlobClient(new Uri(bloburi), new BlobClientOptions());
            
            // prepare query Option
            IProgress<long> progressHandler = new Progress<long>((finishedBytes) =>
            {
                Console.WriteLine(finishedBytes);
            });
            BlobQueryOptions queryOption = new BlobQueryOptions
            {
                InputTextConfiguration = new BlobQueryCsvTextConfiguration() { HasHeaders = true },
                OutputTextConfiguration = new BlobQueryCsvTextConfiguration() { HasHeaders = true },
                ProgressHandler = progressHandler
            };

            string query = "SELECT * FROM BlobStorage WHERE _1 LIKE '1%%'";
            Task<Response<BlobDownloadInfo>>  task =blob.QueryAsync(query, queryOption);
            task.Wait();

            var reader = task.Result.Value.Content;
            FileStream fs = File.Create(@"C:\temp\QA_result.csv");
            reader.CopyTo(fs); 
            fs.Close();
           reader.Close();

```

**Environment:**
 - Name and version of the Library package used: [e.g. Azure.Storage.Blobs 12.4.0] 
Azure.Storage.Blobs/12.5.0-preview.5
 - Hosting platform or OS and .NET runtime version (`dotnet --info` output for .NET Core projects): [e.g. Azure AppService or Windows 10 .NET Framework 4.8]
```
.NET Core SDK (reflecting any global.json):
 Version:   2.2.108
 Commit:    33ed5b90ce

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18363
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.2.108\

Host (useful for support):
  Version: 2.2.6
  Commit:  7dac9b1b51

.NET Core SDKs installed:
  2.2.108 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
```
 - IDE and version : [e.g. Visual Studio 16.3]
VS 15.7.4

### PR DESCRIPTION
**Pull request by seanmcc-msft**
Resolves https://github.com/Azure/azure-sdk-for-net/issues/13441

### CODE REVIEW

**Comment by seanmcc-msft** [comment]
@kasobol-msft @amnguye

File: sdk/storage/Azure.Storage.Blobs/tests/BlobQuickQueryTests.cs [hunk 1]
```diff
@@ -126,14 +126,14 @@ public async Task QueryAsync_MultipleDataRecords()
         }
 
         [Test]
-        [Ignore("Don't want to record 120 MB of data.")]
+        [Ignore("Don't want to record 250 MB of data.")]
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs/tests/BlobQuickQueryTests.cs:129]: Should this be live test ?

**Review by kasobol-msft — APPROVED** [review]

File: sdk/storage/Azure.Storage.Blobs/tests/BlobQuickQueryTests.cs (hunk 1 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs/tests/BlobQuickQueryTests.cs:129]: I think it will timeout and/or be flaky.

Comment (kasobol-msft) [reply sdk/storage/Azure.Storage.Blobs/tests/BlobQuickQueryTests.cs:129]: We can give it a try and see what happens.

**Review by amnguye — APPROVED** [review]

File: sdk/storage/Azure.Storage.Blobs/src/BlobQuickQueryStream.cs [hunk 2]
```diff
@@ -131,6 +131,14 @@ private async Task<int> ReadInternal(bool async, byte[] buffer, int offset, int
                     case Constants.QuickQuery.DataRecordName:
                         record.TryGetValue(Constants.QuickQuery.Data, out object byteObject);
                         byte[] bytes = (byte[])byteObject;
+
+                        if (_buffer != null)
+                        {
+                            ArrayPool<byte>.Shared.Return(_buffer);
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Blobs/src/BlobQuickQueryStream.cs:137]: Always set `clearArray = true` when there's customer data

Comment (tg-msft) [reply sdk/storage/Azure.Storage.Blobs/src/BlobQuickQueryStream.cs:137]: Nit - we could check whether the current buffer size is larger than `bytes.Length` and not bother returning/renting a new one.

File: sdk/storage/Azure.Storage.Blobs/src/BlobQuickQueryStream.cs [hunk 3]
```diff
@@ -260,6 +268,11 @@ public override long Position
         protected override void Dispose(bool disposing)
         {
             _avroStream.Dispose();
+            if (_buffer != null)
+            {
+                ArrayPool<byte>.Shared.Return(_buffer);
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Blobs/src/BlobQuickQueryStream.cs:273]: Pass `clearArray = true` here as well

File: sdk/storage/Azure.Storage.Blobs/tests/BlobQuickQueryTests.cs (hunk 1 above)
Comment (tg-msft) [reply sdk/storage/Azure.Storage.Blobs/tests/BlobQuickQueryTests.cs:129]: We should never use `[Ignore]` without a work item.  If we're not planning to make the test stable someday, we might as well delete it now.

File: sdk/storage/Azure.Storage.Blobs/src/BlobQuickQueryStream.cs [hunk 4]
```diff
@@ -131,6 +131,14 @@ private async Task<int> ReadInternal(bool async, byte[] buffer, int offset, int
                     case Constants.QuickQuery.DataRecordName:
                         record.TryGetValue(Constants.QuickQuery.Data, out object byteObject);
                         byte[] bytes = (byte[])byteObject;
+
+                        if (_buffer != null)
+                        {
+                            ArrayPool<byte>.Shared.Return(_buffer);
+                        }
+
+                        _buffer = ArrayPool<byte>.Shared.Rent(bytes.Length);
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Blobs/src/BlobQuickQueryStream.cs:140]: Consider doing `Math.Max(4MB, bytes.Length)` to rent a buffer that'll be big enough for most records

**Review by tg-msft — APPROVED** [review]
So the bug was hitting a payload larger than 4MB?  The fix makes sense and looks good.

File: sdk/storage/Azure.Storage.Blobs/src/BlobQuickQueryStream.cs (hunk 3 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs/src/BlobQuickQueryStream.cs:273]: Fixed in next iteration.

File: sdk/storage/Azure.Storage.Blobs/src/BlobQuickQueryStream.cs (hunk 2 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs/src/BlobQuickQueryStream.cs:137]: Fixed in next iteration.

File: sdk/storage/Azure.Storage.Blobs/src/BlobQuickQueryStream.cs (hunk 4 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs/src/BlobQuickQueryStream.cs:140]: Fixed in next iteration.

File: sdk/storage/Azure.Storage.Blobs/tests/BlobQuickQueryTests.cs (hunk 1 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs/tests/BlobQuickQueryTests.cs:129]: This test use useful for local testing, to make sure we didn't break the large query cases.

Comment (kasobol-msft) [reply sdk/storage/Azure.Storage.Blobs/tests/BlobQuickQueryTests.cs:129]: Maybe we should spin up additional test project that's run separately from nightly tests - i.e. before release and locally on demand. We've been struggling to find a good home for resource heavy tests like that for a while.
Also our libraries promise to handle large payloads well, so we should have an automated way to verify this aspect. If infra is a problem here let's invest in infra.

**Pull request merged**